jdbc/src/main/java/nextstep/jdbc/JdbcTemplate.java

```diff
@@ -1,9 +1,91 @@
 package nextstep.jdbc;
 
+import com.google.common.base.Strings;
+import nextstep.jdbc.exception.IncorrectResultSizeDataAccessException;
+import nextstep.jdbc.exception.InvalidQueryParameterException;
+import nextstep.jdbc.exception.JdbcTemplateSqlException;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import javax.sql.DataSource;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.IntStream;
+
 public class JdbcTemplate {
 
     private static final Logger log = LoggerFactory.getLogger(JdbcTemplate.class);
+
+    private final DataSource dataSource;
+
+    public JdbcTemplate(DataSource dataSource) {
+        this.dataSource = dataSource;
+    }
+
+    public <T> T queryForObject(String sql, RowMapper<T> rowMapper, Object... objects) {
+        List<T> results = query(sql, rowMapper, objects);
+        if (results.isEmpty()) {
+            return null;
+        }
+
+        if (results.size() > 1) {
+            throw new IncorrectResultSizeDataAccessException(1, results.size());
+        }
+
+        return results.iterator().next();
+    }
+
+    public <T> List<T> query(String sql, RowMapper<T> rowMapper, Object... objects) {
+        assert !Strings.isNullOrEmpty(sql) : "Query is Null And Empty!";
```

Comment:
매의 눈!!!!!!! 쿼리가 제대로 들어왔는지도 체크하시다니 꼼꼼하시군요..! 👍👍

그런데 assertion이 실서버에서는 동작하지 않을 수 있어서 Exception을 활용하는 걸 추천한다는데, 이에 대해선 어떻게 생각하시나요?

```suggestion
        assert !Strings.isNullOrEmpty(sql) : "Query is Null Or Empty!";
```

Reply:
12년도에 작성된 글이라 지금도 동일할지 모르겠지만 파피가 코멘트 주신 내용은 [해당 참고글](https://blog.xenomity.com/Java-Assertion%EC%9D%84-%EC%82%AC%EC%9A%A9%ED%95%98%EC%A7%80-%EB%A7%90%EC%9E%90/)의 내용이 아닐까 싶어요. 

전 평소에 프로덕션 코드에 `assert` 구문을 안써서 이런 부분은 전혀 몰랐네요!  
참고자료의 내용들을 봤을 때, 프로덕션 코드에 assertion 사용은 가급적 피하는게 좋을 것 같네요 🤔